[EV-1090] - [NODE] Back-End Criação do usuário

djenifergurskas opened the pull request:
Requisições do HTTP;
Manipulação de rotas;
Validação dos campos do usuário conforme a tipagem;
Retorno da validação dos campos;
Insomnia utilizado nos testes unitários;

djenifergurskas committed:
commit

commit

djenifergurskas committed:
commit

commit

djenifergurskas committed:
feat

add router patch in routes.

ArtemisNeoma commented on src/index.js:
```diff
@@ -1,11 +1,42 @@
-app.use(routes)
+app.use(router)
+
+app.get("/users", (req, res) => {
```
Essa rota não é um tanto redundante? Dado que o RoutesCustomers deveria ser o único responsável pelos customers e já possui uma rota com GET que retorna customers

djenifergurskas replied:
Sim, tens razão. Já corrigi! 👍

ArtemisNeoma commented on src/index.js:
```diff
@@ -1,11 +1,42 @@
+  res.json(customers);
+});
+
+app.post("/create_user", customerValidator, (req, res,) => {
```
Aqui também deveria estar com o RoutesCustomer dado que esse é o responsável, além de seguir de maneira mais próxima aquilo descrito no card.

djenifergurskas replied:
Certo, já consegui arrumar. 👍

ArtemisNeoma commented on src/index.js:
```diff
@@ -1,11 +1,42 @@
+        }
+        var listCustomer = new CustomerModel(req.body);
+        customers.push(listCustomer);
+        res.status(200).json({ status: "success", customer_created: listCustomer });
```
O código http aqui deveria ser 201 (CREATED) ^^

djenifergurskas replied:
Sim! Eu não tenho muita noção dos códigos de http, tenho que melhorar isto, obrigada! Consegui corrigir.

ArtemisNeoma commented on src/index.js:
```diff
@@ -1,11 +1,42 @@
+        
+})
+
+app.put("/update", (req, res) => {
```
Não é preciso definir o comportamento da rota para todos as ações. Nesse caso como apenas é preciso do POST e GET, apenas eles você precisa definir.

djenifergurskas replied:
Certo!

ArtemisNeoma commented on src/routes/RoutesCustomer.js:
```diff
@@ -0,0 +1,26 @@
+  res.json(customers);
+});
+
+router.put("/", (req, res) => {
```
Não é preciso definir o comportamento da rota para todos as ações.

ismaelalvesgit replied:
Se o objeito da task e somente a criação do usuario acredito que os metodos

- get
- patch
- delete
- put

deveram ser removidos

djenifergurskas replied:
Sim. As rotas já foram removidas, obrigada pessoal!

ArtemisNeoma commented on package.json:
```diff
@@ -8,7 +8,10 @@
   "license": "MIT",
   "dependencies": {
-    "express": "^4.17.3",
+    "cpf-cnpj-validator": "^1.0.3",
```
Essa biblioteca está sendo usada em algum local?

djenifergurskas replied:
Eu iria utilizar para uma validação no CPF, mas no final acabei nem utilizando.

ismaelalvesgit commented on src/index.js:
```diff
@@ -1,11 +1,42 @@
+        if (!errors.isEmpty()) {
+            return res.status(422).json({ errors: errors.mapped() })
+        }
+        var listCustomer = new CustomerModel(req.body);
```
não user o verbo `var` para criar variaveis por favor adote `let` ou `const`

djenifergurskas replied:
Certo!

ismaelalvesgit commented on package.json:
```diff
@@ -8,7 +8,10 @@
-    "express": "^4.17.3",
+    "cpf-cnpj-validator": "^1.0.3",
+    "express": "^4.18.2",
+    "express-validator": "^6.14.2",
```
Recomendo tambem que deixe de usar o [express-validador](https://express-validator.github.io/docs) pois a forma de como e feita a validação dessa lib alem de ser muito verboso diminuei a capacidade de seu uso na api limitando somente a estrutura do express. recomendo que use a lib [joi](https://joi.dev/api/?v=17.7.0) pois lhe dara mais flexibilidade.

djenifergurskas replied:
Ok!

ismaelalvesgit commented on src/index.js:
```diff
@@ -1,11 +1,42 @@
+});
+
+app.post("/create_user", customerValidator, (req, res,) => {
+  const errors = validationResult(req)
```
Acho que esse trecho de codigo deveria virar um middleware que fosse resposanvel por realizar as validações de requesição recebendo como exemplo um paramentro contendo as informações a serem validadas.

ex:

```ts
const validator =  (schema: Validation)=>{

    return (req, res, next)=>{
        // logica de validação
    }
}

// uso
router.get('/user', validator(schema_a_ser_validado), controller)
```

djenifergurskas replied:
Certo!

djenifergurskas committed:
refactor

mudanças nos arquivos da rotas, remoção de rotas não utilizadas e remoção da biblioteca cpf-cnpj-validator que não estava sendo utilizada.
